File: decode/go/imagedecode/gif.go
package imagedecode

import (
	"bytes"
	"image"
	"image/gif"
)

// gifGlobalColorTable reads the global color table from raw GIF bytes, returning
// the RGB triples (one [3]byte per palette entry). Returns nil if no global
// color table is present or the data is too short.
func gifGlobalColorTable(b []byte) [][3]byte {
	if len(b) < 13 {
		return nil
	}
	flags := b[10]
	if flags>>7 == 0 {
		return nil // no global color table
	}
	entries := 1 << (int(flags&0x7) + 1)
	offset := 13
	if offset+entries*3 > len(b) {
		return nil
	}
	palette := make([][3]byte, entries)
	for i := 0; i < entries; i++ {
		palette[i] = [3]byte{b[offset], b[offset+1], b[offset+2]}
		offset += 3
	}
	return palette
}

// gifTransparentIndex scans the GIF byte stream for a Graphic Control Extension
// and returns the transparent color index, or -1 if none.
func gifTransparentIndex(b []byte, gctEntries int) int {
	pos := 13 + gctEntries*3
	for pos < len(b) {
		switch b[pos] {
		case 0x21: // Extension introducer
			if pos+1 >= len(b) {
				return -1
			}
			isGCE := b[pos+1] == 0xF9
			pos += 2
			for pos < len(b) {
				sz := int(b[pos])
				pos++
				if sz == 0 {
					break
				}
				if isGCE && sz >= 4 && pos+sz-1 < len(b) {
					packed := b[pos]
					if packed&0x01 != 0 {
						return int(b[pos+3])
					}
					isGCE = false
				}
				pos += sz
			}
		case 0x2C: // Image descriptor
			return -1
		case 0x3B: // Trailer
			return -1
		default:
			pos++
		}
	}
	return -1
}

func decodeGif(b []byte) (DecodedImage, error) {
	img, err := gif.Decode(bytes.NewReader(b))
	if err != nil {
		return DecodedImage{}, newError(CorruptInput, Gif, true, "gif.Decode failed: "+err.Error())
	}
	bnds := img.Bounds()
	width := bnds.Dx()
	height := bnds.Dy()

	p, isPaletted := img.(*image.Paletted)

	// Determine if the image has transparency and which palette index is transparent.
	// Go's image/gif decoder zeroes the RGB of the transparent palette entry (sets it to
	// {0,0,0,0}). PIL/Pillow preserves the original palette RGB. To match the reference
	// golden outputs we read the original palette from the raw GIF bytes.
	transpIdx := -1
	var rawPalette [][3]byte
	if isPaletted {
		rawPalette = gifGlobalColorTable(b)
		if rawPalette != nil {
			transpIdx = gifTransparentIndex(b, len(rawPalette))
		}
	}

	hasAlpha := transpIdx >= 0
	if !hasAlpha && !isPaletted {
		// Non-palette GIF (unusual): scan all pixels for alpha != 0xFFFF
		for y := bnds.Min.Y; y < bnds.Max.Y && !hasAlpha; y++ {
			for x := bnds.Min.X; x < bnds.Max.X; x++ {
				_, _, _, a := img.At(x, y).RGBA()
				if a != 0xFFFF {
					hasAlpha = true
					break
				}
			}
		}
	}

	channelCount := 3
	channels := RGB
	if hasAlpha {
		channelCount = 4
		channels = RGBA
	}

	out := make([]byte, width*height*channelCount)
	idx := 0

	if isPaletted && rawPalette != nil {
		// Fast path: use raw palette index to avoid premultiplied-alpha loss.
		// For transparent pixels, use the original palette RGB (not Go's zeroed {0,0,0}).
		for y := bnds.Min.Y; y < bnds.Max.Y; y++ {
			for x := bnds.Min.X; x < bnds.Max.X; x++ {
				pixOffset := p.PixOffset(x, y)
				palIdx := int(p.Pix[pixOffset])
				var r, g, bl byte
				var a byte = 255
				if palIdx < len(rawPalette) {
					rgb := rawPalette[palIdx]
					r, g, bl = rgb[0], rgb[1], rgb[2]
				}
				if palIdx == transpIdx {
					a = 0
				}
				out[idx] = r
				out[idx+1] = g
				out[idx+2] = bl
				idx += 3
				if hasAlpha {
					out[idx] = a
					idx++
				}
			}
		}
	} else {
		// Fallback: use At() for non-paletted images.
		for y := bnds.Min.Y; y < bnds.Max.Y; y++ {
			for x := bnds.Min.X; x < bnds.Max.X; x++ {
				r, g, bl, a := img.At(x, y).RGBA()
				out[idx] = byte(r >> 8)
				out[idx+1] = byte(g >> 8)
				out[idx+2] = byte(bl >> 8)
				idx += 3
				if hasAlpha {
					out[idx] = byte(a >> 8)
					idx++
				}
			}
		}
	}

	return DecodedImage{Width: width, Height: height, Data: out, Channels: channels, Format: Gif}, nil
}

File: decode/rust/rosetta-image-decode/src/gif.rs
use crate::error::{DecodeError, DecodeErrorKind};
use crate::limits::check_dimensions;
use crate::types::{Channels, DecodedImage, Format};

use image::ImageReader;
use std::io::Cursor;

/// Returns true if the GIF's first frame has a Graphic Control Extension
/// with the transparency flag set — matching PIL's img.info['transparency'] logic.
fn gif_has_transparency(bytes: &[u8]) -> bool {
    if bytes.len() < 13 {
        return false;
    }
    // Skip 6-byte header + 7-byte logical screen descriptor
    let packed = bytes[10];
    let gct_flag = (packed >> 7) & 1;
    let gct_size = packed & 7;

    let mut i: usize = 13;
    // Skip global color table
    if gct_flag != 0 {
        let n_colors = 1usize << (gct_size + 1);
        i += n_colors * 3;
    }

    // Walk blocks until we hit the first image descriptor
    while i < bytes.len() {
        let block_type = bytes[i];
        match block_type {
            0x3B => return false, // GIF trailer
            0x21 => {
                // Extension block
                if i + 1 >= bytes.len() {
                    return false;
                }
                let ext_label = bytes[i + 1];
                if ext_label == 0xF9 {
                    // Graphic Control Extension
                    // Format: 0x21 0xF9 <block-size=4> <flags> <delay-lo> <delay-hi> <transparent-idx> 0x00
                    if i + 3 < bytes.len() {
                        let flags = bytes[i + 3];
                        if flags & 1 != 0 {
                            return true;
                        }
                    }
                    // This GCE belongs to the first image — if it has no transparency, stop.
                    // Skip extension sub-blocks
                    i += 2;
                    while i < bytes.len() {
                        let sub_len = bytes[i] as usize;
                        i += 1;
                        if sub_len == 0 {
                            break;
                        }
                        i += sub_len;
                    }
                    // After the GCE we expect the image descriptor (0x2C), don't keep scanning
                    return false;
                } else {
                    // Other extension — skip it
                    i += 2;
                    while i < bytes.len() {
                        let sub_len = bytes[i] as usize;
                        i += 1;
                        if sub_len == 0 {
                            break;
                        }
                        i += sub_len;
                    }
                }
            }
            0x2C => {
                // Image descriptor — first frame reached, no GCE transparency found
                return false;
            }
            _ => {
                // Unknown block, stop
                return false;
            }
        }
    }
    false
}

pub(crate) fn decode_gif(bytes: &[u8]) -> Result<DecodedImage, DecodeError> {
    let reader = ImageReader::with_format(Cursor::new(bytes), image::ImageFormat::Gif);
    let img = match reader.decode() {
        Ok(i) => i,
        Err(e) => {
            return Err(DecodeError::new(
                DecodeErrorKind::CorruptInput,
                Some(Format::Gif),
                format!("image::decode failed: {}", e),
            ))
        }
    };

    let (width, height) = (img.width() as usize, img.height() as usize);
    check_dimensions(width, height, Format::Gif)?;

    // The image crate always returns RGBA8 for GIF palette images.
    // We replicate PIL's behaviour: use RGBA only if the first frame has a
    // Graphic Control Extension with the transparency flag set.
    if gif_has_transparency(bytes) {
        let data: Vec<u8> = img.to_rgba8().into_raw();
        Ok(DecodedImage {
            width,
            height,
            data,
            channels: Channels::Rgba,
            format: Format::Gif,
        })
    } else {
        let data: Vec<u8> = img.to_rgb8().into_raw();
        Ok(DecodedImage {
            width,
            height,
            data,
            channels: Channels::Rgb,
            format: Format::Gif,
        })
    }
}
